_posts/2022-08-08-understanding-meta-keyword-argument.md

```diff
@@ -0,0 +1,240 @@
+---
+layout: post
+title: Understanding Dask’s meta keyword argument
+author: Pavithra Eswaramoorthy
+tags: [dataframe]
+theme: twitter
+---
+
+{% include JB/setup %}
+
+If you have worked with Dask DataFrames or Dask Arrays, you have probably come across the `meta` keyword argument and seen one or more of the following warning/error messages:
+
+```
+UserWarning: You did not provide metadata, so Dask is running your function on a small dataset to guess output types. It is possible that Dask will guess incorrectly.
+```
+
+```
+UserWarning: `meta` is not specified, inferred from partial data. Please provide `meta` if the result is unexpected.
+```
+
+```
+ValueError: Metadata inference failed in …
+```
+
+In this blog post, we will discuss:
+
+- what the is `meta` keyword argument, and
+- how to use `meta` effectively
+
+with some minimal examples.
+
+We will look at `meta` mainly in the context of Dask DataFrames, however, similar principles also apply to Dask Arrays.
+
+## What is `meta`?
+
+Before answering this, let's quickly discuss Dask DataFrames.
+
+A Dask DataFrame is a lazy object composed of multiple pandas DataFrames, where each pandas DataFrame is called a "partition". These are stacked along the index and Dask keeps track of these partitions using “divisions”, which is a tuple representing the start and end index of each partition.
+
+![Dask DataFrame consists of multiple pandas DataFrames](https://docs.dask.org/en/stable/_images/dask-dataframe.svg)
+
+When you create a Dask DataFrame, you usually see something like the following:
+
+```python
+>>> import pandas as pd
+>>> import dask.dataframe as dd
+>>> ddf = dd.DataFrame.from_dict(
+...     {
+...         "x": range(6),
+...         "y": range(10, 16),
+...     },
+...     npartitions=2,
+... )
+>>> ddf
+Dask DataFrame Structure:
+                   x      y
+npartitions=2
+0              int64  int64
+3                ...    ...
+5                ...    ...
+Dask Name: from_pandas, 2 tasks
+```
+
+Here, Dask has created the structure of the DataFrame using some "metadata" information about the _column names_ and their _datatypes_. Dask uses this metadata for understanding Dask operations and creating accurate task graphs (i.e., the logic of your computation).
+
+Internally, this metadata is represented as an empty pandas DataFrame or Series, which has the same structure as your output Dask DataFrame. To learn more about how `meta` is defined internally, check out the [DataFrame Internal Design documentation](https://docs.dask.org/en/stable/dataframe-design.html#metadata).
+
+The `meta` keyword argument in various Dask DataFrame functions allows you to explicitly share this metadata information with Dask. Note that the keyword argument is concerned with the metadata of the _output_ of those functions.
+
+To see the actual metadata information for a collection, you can look at the `._meta` attribute:
```

Comment:
_Not a blocker_. Should we think about exposing this via a public method in dask/dask? Maybe with a method like `get_meta()` or `view_meta()`? Getting folks to look at a private property feels unusual.

Reply:
Good point! Julia already has an issue open about this: https://github.com/dask/dask/issues/8585 :)

I agree that `._meta` is not the best way to go. I included it because it's pretty useful, even for high-level users. Maybe we can keep it until we have the public methods? I'll add a link to Julia's issue and a disclaimer about `._meta` being private. I'll also make sure to update this post when we do have the public methods. How does that sound?

Reply:
Sounds great!